packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -1254,14 +1254,18 @@ class _SliverAppBarDelegate extends SliverPersistentHeaderDelegate {
         child: title,
       ),
     };
+    final ScaffoldState? scaffold = Scaffold.maybeOf(context);
+    final ModalRoute<dynamic>? parentRoute = ModalRoute.of(context);
+    final bool hasLeading = leading != null || (scaffold?.hasDrawer ?? false)
+      || (parentRoute?.impliesAppBarDismissal ?? false);
 
     final Widget appBar = FlexibleSpaceBar.createSettings(
       minExtent: minExtent,
       maxExtent: maxExtent,
       currentExtent: math.max(minExtent, maxExtent - shrinkOffset),
       toolbarOpacity: toolbarOpacity,
       isScrolledUnder: isScrolledUnder,
-      hasLeading: leading != null || automaticallyImplyLeading,
+      hasLeading: hasLeading,
```

Comment:
Should automaticallyImplyLeading still have some effect here?

Reply:
Great catch. 
We should check for `hasDrawer` and `impliesAppBarDismissal` if `leading` is `null` and `automaticallyImplyLeading` is true.

Reply:
Updated with the following
```dart
    if (leading == null && automaticallyImplyLeading) {
      hasLeading = (scaffold?.hasDrawer ?? false) || (parentRoute?.impliesAppBarDismissal ?? false);
    }
```